File: extraction-service/services/path_c/sld_topology_service.py
import logging

from models.response import CubicleDeviceResult, TopologySummary

logger = logging.getLogger(__name__)

PANEL_TYPE_KEYWORDS: dict[str, list[str]] = {
    "Incomer":     ["arriv", "incom", "arrivée", "arrivo", "einspeise", "incoming", "source"],
    "Feeder":      ["départ", "feeder", "sortie", "partenza", "abgang", "outgoing", "load"],
    "Coupler":     ["couplage", "coupler", "bus-tie", "accoppiatore", "kupplung", "bustie", "tie"],
    "Metering":    ["comptage", "metering", "mesure", "misura", "messung", "meter"],
    "Transformer": ["transfo", "transformer", "trasformatore", "transformateur", "trafo"],
    "VT":          ["vt panel", "voltage transformer", "pt panel", "transformateur de tension",
                    "trasformatore tensione", "spannungswandler", "vt feeder"],
    "BusRiser":    ["remontée", "bus riser", "risalita", "riser"],
}


def classify_panel_type(text: str) -> str:
    lower = text.lower()
    for panel_type, keywords in PANEL_TYPE_KEYWORDS.items():
        if any(kw in lower for kw in keywords):
            return panel_type
    return "Unknown"


def _deduplicate(devices: list[CubicleDeviceResult]) -> list[CubicleDeviceResult]:
    """
    Keep the highest-confidence device when multiple paths report the same
    functional_position on the same source page.
    """
    seen: dict[tuple[str, int], CubicleDeviceResult] = {}
    for d in devices:
        key = (d.functional_position.strip().lower(), d.source_page)
        existing = seen.get(key)
        if existing is None or d.confidence > existing.confidence:
            seen[key] = d
    return list(seen.values())


def build_topology(devices: list[CubicleDeviceResult]) -> TopologySummary:
    # Reclassify Unknown panels without mutating the originals
    reclassified: list[CubicleDeviceResult] = []
    for d in devices:
        if not d.panel_type or d.panel_type == "Unknown":
            guessed = classify_panel_type(d.functional_position)
            if guessed != "Unknown":
                d = d.model_copy(update={"panel_type": guessed})
        reclassified.append(d)

    # Deduplicate before counting — prevents double-counting when
    # Path B and Path C both extract the same device
    unique = _deduplicate(reclassified)

    counts = {
        t: sum(1 for d in unique if d.panel_type == t)
        for t in ["Incomer", "Feeder", "Coupler", "Metering", "Transformer", "VT"]
    }

    # Each coupler splits the busbar into one additional section
    busbar_sections = max(1, counts["Coupler"] + 1)

    parts = [f"{counts['Incomer']} incomer(s)"]
    if counts["Coupler"]:
        parts.append(f"{counts['Coupler']} coupler(s)")
    parts.append(f"{counts['Feeder']} feeder(s)")
    if counts["Metering"]:
        parts.append(f"{counts['Metering']} metering panel(s)")
    if counts["Transformer"]:
        parts.append(f"{counts['Transformer']} transformer(s)")
    if counts["VT"]:
        parts.append(f"{counts['VT']} VT panel(s)")
    parts.append(f"{busbar_sections} busbar section(s)")
    desc = " + ".join(parts)

    logger.info(f"[PathC-SF3] Topology: {desc} (from {len(unique)} unique panels)")

    return TopologySummary(
        total_panels=len(unique),
        incomers=counts["Incomer"],
        feeders=counts["Feeder"],
        couplers=counts["Coupler"],
        metering=counts["Metering"],
        transformers=counts["Transformer"],
        busbar_sections=busbar_sections,
        description=desc,
    )

File: backend/SwitchgearApi/Services/SldTopologyService.cs
using System.Linq;
using Microsoft.Extensions.Logging;

namespace SwitchgearApi.Services
{
    public interface ISldTopologyService
    {
        PythonTopologySummary BuildTopology(SldVisionResult sldResult);
    }

    public class SldTopologyService : ISldTopologyService
    {
        private readonly ILogger<SldTopologyService> _logger;

        public SldTopologyService(ILogger<SldTopologyService> logger)
        {
            _logger = logger;
        }

        public PythonTopologySummary BuildTopology(SldVisionResult sldResult)
        {
            var summary = new PythonTopologySummary();

            foreach (var panel in sldResult.Panels)
            {
                summary.TotalPanels++;
                var type = panel.Type ?? string.Empty;
                var desc = panel.Description ?? string.Empty;

                if (IsIncomer(type, desc))        summary.Incomers++;
                else if (IsFeeder(type, desc))    summary.Feeders++;
                else if (IsCoupler(type, desc))   summary.Couplers++;
                else if (IsMetering(type, desc))  summary.Metering++;
                else if (IsTransformer(desc))     summary.Transformers++;
            }

            summary.BusbarSections = summary.Couplers + 1;

            summary.Description = BuildDescription(summary);

            _logger.LogInformation(
                "[Topology] {TotalPanels} panels → {Description}",
                summary.TotalPanels, summary.Description);

            return summary;
        }

        private static bool IsIncomer(string type, string desc) =>
            Eq(type, "incomer") ||
            ContainsAny(desc, "arriv", "incom", "arrivée", "arrivo", "einspeise");

        private static bool IsFeeder(string type, string desc) =>
            Eq(type, "feeder") ||
            ContainsAny(desc, "départ", "feeder", "sortie", "partenza", "abgang");

        private static bool IsCoupler(string type, string desc) =>
            Eq(type, "coupler") ||
            ContainsAny(desc, "couplage", "coupler", "bus-tie", "accoppiatore", "kupplung");

        private static bool IsMetering(string type, string desc) =>
            Eq(type, "measurement") ||
            ContainsAny(desc, "comptage", "metering", "mesure", "misura", "messung");

        private static bool IsTransformer(string desc) =>
            ContainsAny(desc, "transfo", "transformer", "trasformatore");

        private static bool Eq(string a, string b) =>
            string.Equals(a, b, System.StringComparison.OrdinalIgnoreCase);

        private static bool ContainsAny(string text, params string[] keywords) =>
            keywords.Any(k => text.Contains(k, System.StringComparison.OrdinalIgnoreCase));

        private static string BuildDescription(PythonTopologySummary s)
        {
            var parts = new System.Collections.Generic.List<string>();
            if (s.Incomers     > 0) parts.Add($"{s.Incomers}-incomer");
            if (s.Feeders      > 0) parts.Add($"{s.Feeders}-feeder");
            if (s.Couplers     > 0) parts.Add($"{s.Couplers}-coupler");
            if (s.Metering     > 0) parts.Add($"{s.Metering}-metering");
            if (s.Transformers > 0) parts.Add($"{s.Transformers}-transformer");

            var joined = parts.Count > 0 ? string.Join(" + ", parts) : "unknown";
            return $"{joined} ({s.BusbarSections} section(s))";
        }
    }
}
